imap/http_jmap.c

```diff
@@ -1170,45 +1177,201 @@ static int jmap_get_session(struct transaction_t *txn)
+                "accountid=<%s>", jpush->accountid);
+        return ev;
+    }
+
+    /* See if anything has changed */
+    json_t *changed = json_object();
+    const struct datatype_t *dtype;
+    for (dtype = dataTypes; dtype->name; dtype++) {
+        modseq_t *modseq = (modseq_t *)((size_t) &jpush->counters + dtype->offset);
+        modseq_t *cur_modseq = (modseq_t *)((size_t) &cur_counters + dtype->offset);
+
+        if (*modseq < *cur_modseq) {
+            *modseq = *cur_modseq;
+
+            buf_reset(&buf);
+            buf_printf(&buf, MODSEQ_FMT, *modseq);
+            json_object_set_new(changed, dtype->name, json_string(buf_cstring(&buf)));
+        }
+    }
+
+    if (json_object_size(changed)) {
+        buf_reset(&buf);
+        buf_printf(&buf, MODSEQ_FMT, cur_counters.highestmodseq);
+
+        json_t *jstate = json_pack("{ s:s s:{ s:o } s:s }",
+                                   "@type", "StateChange",
+                                   "changed", jpush->accountid, changed,
+                                   "pushState", buf_cstring(&buf));
+
+        /* Send the StateChange object */
+        size_t flags = JSON_PRESERVE_ORDER |
+            (config_httpprettytelemetry ? JSON_INDENT(2) : JSON_COMPACT);
+        buf_initmcstr(&buf, json_dumps(jstate, flags));
+        ws_send(txn, &buf);
+        prot_flush(txn->conn->pout);
+
+        json_decref(jstate);
+    }
+    else {
+        json_decref(changed);
+    }
+
+    buf_free(&buf);
+    
+    /* Schedule our next update */
+    ev->mark = time(NULL) + push_poll;
+
+    return ev;
+}
+
+static int jmap_ws_push_enable(struct transaction_t *txn, json_t *req)
+{
+    struct ws_context *ctx = (struct ws_context *) txn->ws_ctx;
+    struct ws_push_ctx *jpush = ws_get_app_data(ctx);
+    json_t *types = req ? json_object_get(req, "dataTypes") : NULL;
+
+    xsyslog(LOG_DEBUG, "JMAP WS push enable",
+            "jpush=<%p>, types=<%p>", jpush, types);
+
+    if (json_is_array(types)) {
+        /* Enable */
+        json_t *pushState = json_object_get(req, "pushState");
+        struct mboxname_counters cur_counters;
+        modseq_t lastmodseq = ULLONG_MAX;
+
+        if (!jpush) {
+            jpush = xzmalloc(sizeof(struct ws_push_ctx));
+            jpush->accountid = xstrdup(httpd_userid);
+            jpush->inboxname = mboxname_user_mbox(jpush->accountid, NULL);
+            ws_set_app_data(ctx, jpush);
+        }
+
+        if (json_is_string(pushState)) {
+            lastmodseq = atomodseq_t(json_string_value(pushState));
+        }
+        else if (mboxname_read_counters(jpush->inboxname, &cur_counters)) {
+            /* Something went wrong */
+            goto disable;
+        }
+
+        /* Initialize our tracking modseqs to the maximum value */
+        const struct datatype_t *dtype;
+        for (dtype = dataTypes; dtype->name; dtype++) {
+            modseq_t *modseq =
+                (modseq_t *)((size_t) &jpush->counters + dtype->offset);
```

Comment:
nit: I think the `(size_t)` cast here (and in the other places that this pattern occurs) should be `(uintptr_t)`, since it looks like the intention is "make sure the addition result is the right type to then be cast to a pointer"?

I think in practice `size_t` and `uintptr_t` are usually the same, but if we're bothering to cast the intermediate result (to shut up some warning, I assume), we might as well cast it to the right thing...